### PR TITLE
[CURL] Add verifyHost

### PR DESCRIPTION
**Pull request by peak3d**
## Description
Add option to set CURLOPT_SSL_VERIFYHOST using verifyhost=[true/false] in URL header.

## Motivation and Context
We already support verifypeer (setting CURLOPT_SSL_VERIFYPEER), butthis is not sufficient to connect to servers with self signed certificates. To simulate the curl --insecure option, both PEER and HOST must be set to false.

## How Has This Been Tested?
Access a stream using HTTPS protocol on a server with a self signed certificate.

## Types of change
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

### CODE REVIEW

**Comment by wsnipex** [comment]
just for completeness: this option is for servers with **broken** self-signed certs, where host/domain name and cert subject cn do not match.
A proper self-signed cert with correct hostname works fine with just verifypeer=false

**Comment by enen92** [comment]
Am I right into thinking that by passing this option as an option Kodi/Curl will skip host verification for all calls?
I have a plugin where some streams fail to play because the server certificate is wrong. To "solve" this I added a setting to prefer http connections over https but I don't exactly like the solution. Would it be possible to pass an option to attempt host verification but allow the connection without verifying the host in case of host verification failure?

**Comment by wsnipex** [comment]
> Am I right into thinking that by passing this option as an option Kodi/Curl will skip host verification for all calls?

it should be just for this URL

> Would it be possible to pass an option to attempt host verification but allow the connection without verifying the host in case of host verification failure?

from security POV, that's exactly the same as turning verify off directly

**Comment by enen92** [comment]
Addon includes 30.000 radio stations, it's impossible to know which ones have broken certificates on the server, so I tried first to simply pass verifyPeer to in the stream that is passed to the player (but that didn't work).

For the second part, you're right, it's pretty much the same thing. Ignore my previous comment/request.

**Comment by peak3d** [comment]
This option is ONLY for this URL. I agree that certificates should always be valid, but this is unfortunately not the case. If devs arrive such an issue, they usually workaround this by providing a proxy, and use python urllib (or other solutions like calling curl --insecure from python).
IMO we do not win anything if we keep back options which are part of libCURL, nowdays devs will always find a workaround.

**Comment by lrusak** [comment]
jenkins build this please

**Comment by yol** [comment]
> Addon includes 30.000 radio stations, it's impossible to know which ones have broken certificates on the server

How exactly do those radio stations work? I mean, every other player software also validates certificates and CNs and will be unable to play the streams.

> IMO we do not win anything if we keep back options which are part of libCURL, nowdays devs will always find a workaround.

There is a difference though in that including the option gives it the notion of being officially accepted. Would we accept addons doing dirty workarounds like doing `curl --insecure` exec calls into the repo? (honest question, not rhetorical)

**Comment by enen92** [comment]
> > Addon includes 30.000 radio stations, it's impossible to know which ones have broken certificates on the server
> 
> How exactly do those radio stations work? I mean, every other player software also validates certificates and CNs and will be unable to play the streams.
> 
> > IMO we do not win anything if we keep back options which are part of libCURL, nowdays devs will always find a workaround.
> 
> There is a difference though in that including the option gives it the notion of being officially accepted. Would we accept addons doing dirty workarounds like doing `curl --insecure` exec calls into the repo? (honest question, not rhetorical)

radio.de API returns a list of available streams for each radio when you query it with the station id. I assume that the android app fallback to the second available url if the first fails. In the case of the failing stream (due to broken certificate), the second available stream in the list is the same url but using http instead of https. I blindly set the resolved url for the item passed to the player, so created a setting to prefer http over https to avoid wasting a lot of time around a single stream.

The proper alternative In kodi (to provide the same functionality) was to use a playlist, but that needs a few hacks to be used with setResolvedUrl (and time I didn't want to spend).

So basically I needed a way to provide the player with a stream that might have a broken cert on the server. This PR provides it (at the cost of disabling verification for any stream that I feed to the player from the addon as I don't have a way of verifying that upfront). So thanks!

**Comment by peak3d** [comment]
@yol are we really getting much more insecure by adding VERIFYHOST additionally to the existing VERIFYPEER? And, isn't VERIFYPEER already a security issue?

**Comment by wsnipex** [comment]
I could argue that a legit radio stream has absolutely no reason to use SSL with broken certs. 
If services don't value or don't need security, they should just use plain http. OTOH, if they need encryption, they  have to do it properly.
Any browser will pop a big fat warning and require the user to explicitly accept the risk, but adding this option here will hide the fact from users, because add-on authors will use it blindly just to make their addon work.

Using this option should at least require user opt in of some kind

**Comment by yol** [comment]
@enen92 Hm unfortunate but I see that we cannot do anything about it if the services are broken. Honestly though, if that's how it is, I guess your prefer-http approach is just fine. If you are going to add `|verifypeer=false,verifyhost=false` to all https URLs (since as you said, you don't keep a list of which servers are broken), you can just as well use plain http. Https gets you nothing relevant in that case.

I agree with @wsnipex, preferably this should be a user choice if at all. Motivation for adding `verifypeer` was user request akin to how Firefox etc. allow you to "connect anyway" in case of a broken cert (#13909). Still, @peak3d you are correct that I don't really like it already. I gave in to it because currently, we don't have a user-friendly, reasonable way to add certificates to the cert store that Kodi uses. So there's no real alternative that I could recommend for using servers with self-signed certs. If we had that, in my opinion we could remove `verifypeer`. Now `verifyhost` only fixes completely broken set-ups where the CN does not match, so the same argument does not apply.

@peak3d Can you explain your concrete usage, where you need this option? Maybe this will help us see the problem.

**Comment by peak3d** [comment]
Sure, @yol , issue was that one of our supporter provided a sample DRM stream for testing Wiseplay DRM (chinese Widevine). Unfortunately both url's (license and manifest) failed to fetch because of the here discussed issue. It was much easier to workaround the self signed cert issue from our side, you may know how long such issues take in bigger companies.

I'm done with testing, and I usually never arrive this issue. Means I have not any personal gain with this PR. Because doubts are still there, I think we can close it and continue as before.

But yeah, we could also add a setting, which the user has to confirm if there is such a situation.
Question would be only: should we ask for every URL, or should we ask once for all?

**Comment by yol** [comment]
That makes sense. And I can see that having this option for debugging/testing is beneficial.

I think that even though the user confirm would be cool, of course, it would probably be too complicated to implement (considering cost vs benefit). You can't just open a dialog from `CCurlFile` - or at the very least, you shouldn't :-)

What about adding an advancedsetting to disable verifyhost/verifypeer globally, for testing? Or is that even more prone to uninformed misuse?

**Comment by phunkyfish** [comment]
I think the advanced setting is a good idea. I’m sure it will get misused but at least it’s use would be explicit on the users behalf.

**Comment by wsnipex** [comment]
what about a domain whitelist for verifyhost/peer in advancedsettings?

**Comment by phunkyfish** [comment]
> what about a domain whitelist for verifyhost/peer in advancedsettings?

even better

**Comment by DaveTBlake** [comment]
> Because doubts are still there, I think we can close it and continue as before.

@peak3d want to close this or add alpha2 milestone?

**Comment by ksooo** [comment]
Nothing for v19.

**Review by rzr — APPROVED** [review]

**Comment by rzr** [comment]
This change looks valuable to me, has it been integrated downstream elsewhere ?
